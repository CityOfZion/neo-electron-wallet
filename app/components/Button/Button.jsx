// @flow

import React, { Component } from 'react'
import { omit } from 'lodash'

import styles from './Button.scss'

type Props = {
  renderIcon: ?Function,
  primary: ?boolean,
  secondary: ?boolean,
  children: React.Node
}

class Button extends Component<Props> {
  static defaultProps = {
    primary: true
  }

  render = () => {
    const { primary, secondary, renderIcon } = this.props
    const passDownProps = omit(this.props, 'primary', 'secondary')
    const conditionalStyles = {}
    if (primary) {
      conditionalStyles.buttonStyle = styles.darkButton
      conditionalStyles.iconStyle = styles.lightIcon
    } else if (secondary) {
      conditionalStyles.buttonStyle = styles.lightButton
      conditionalStyles.iconStyle = styles.darkIcon
    }
    const { iconStyle, buttonStyle } = conditionalStyles

    return (
      <button
        className={`${styles.buttonRedesign} ${buttonStyle}`}
        {...passDownProps}
      >
        {renderIcon ? (
          <span className={`${styles.icon} ${iconStyle}`}>{renderIcon()}</span>
        ) : (
          <div className={styles.icon} />
        )}
        <span>{this.props.children}</span>
      </button>
    )
  }
}

export default Button
