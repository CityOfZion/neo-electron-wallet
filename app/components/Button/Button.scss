@import '../../styles/variables';

.button {
  border: none;
  border-radius: 4px;
  width: 100%;
  display: flex;
  flex-direction: row;
  align-items: center;
  font-size: 14px;
  font-weight: 100;
  padding: 10px;
  white-space: nowrap;
  transition: opacity 0.2s;

  &[disabled] {
    opacity: 0.25;
  }

  &:not([disabled]) {
    cursor: pointer;

    &:hover {
      background-image: linear-gradient(to right, #ee6d66, #a866ee),
        radial-gradient(circle at top left, #ee6d66, #a866ee);
    }
  }

  &.dark {
    background: $primary-button;
    color: #fff;

    path {
      fill: #fff;
    }
  }

  &.light {
    border: solid;
    border-width: 2px;
    border-color: $primary-button;
    background: #fff;
    color: $primary-button;

    &:not([disabled]):hover {
      border: double 2px transparent;
      background-origin: border-box;
      background-clip: content-box, border-box;
      border-color: transparent;
      border-color: transparent;
      color: white;

      path {
        fill: #fff;
      }
    }
  }

  span {
    width: 100%;
  }

  .icon {
    svg {
      height: 25px;
      width: 25px;
    }

    height: 25px;
    width: 25px;
    align-self: flex-start;

    &.dark {
      // TODO
    }

    &.light {
      // TODO
    }
  }
}
