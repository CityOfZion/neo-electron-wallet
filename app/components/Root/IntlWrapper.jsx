// @flow
import React from 'react'
import { compose } from 'recompose'
import '@formatjs/intl-pluralrules/polyfill'
import '@formatjs/intl-pluralrules/polyfill-locales'
import { IntlProvider } from 'react-intl'

import translations from '../../translations'
import withLanguageData from '../../hocs/withLanguageData'
import { LANGUAGES } from '../../core/constants'

const { english, korean, french } = translations

type IntlWrapperProps = {
  children: any,
  language: string,
}

const tranlationsMappings = {
  [LANGUAGES.ENGLISH.value]: english,
  [LANGUAGES.KOREAN.value]: korean,
  [LANGUAGES.FRENCH.value]: french,
}

class IntlWrapper extends React.Component<IntlWrapperProps> {
  render() {
    const { children, language } = this.props
    return (
      <IntlProvider
        locale="en"
        messages={tranlationsMappings[language] || english}
      >
        {children}
      </IntlProvider>
    )
  }
}

export default compose(withLanguageData())(IntlWrapper)
