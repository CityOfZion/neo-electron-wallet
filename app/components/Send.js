import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { doSendAsset, verifyAddress } from 'neon-js'
import { togglePane } from '../modules/dashboard'
import { sendEvent, clearTransactionEvent, toggleAsset } from '../modules/transactions'
import SplitPane from 'react-split-pane'
import ReactTooltip from 'react-tooltip'
import { log } from '../util/Logs'
import { ledgerNanoS_CreateSignature,ledgerNanoS_doSendAsset } from '../modules/ledgerNanoS'

let sendAddress, sendAmount, confirmButton

// form validators for input fields
const validateForm = (dispatch, neoBalance, gasBalance, asset) => {
  // check for valid address
  try {
    if (verifyAddress(sendAddress.value) !== true || sendAddress.value.charAt(0) !== 'A') {
      dispatch(sendEvent(false, 'The address you entered was not valid.'))
      setTimeout(() => dispatch(clearTransactionEvent()), 5000)
      return false
    }
  } catch (e) {
    dispatch(sendEvent(false, 'The address you entered was not valid.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  }
  if (asset === 'Neo' && parseFloat(sendAmount.value) !== parseInt(sendAmount.value)) { // check for fractional neo
    dispatch(sendEvent(false, 'You cannot send fractional amounts of Neo.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  } else if (asset === 'Neo' && parseInt(sendAmount.value) > neoBalance) { // check for value greater than account balance
    dispatch(sendEvent(false, 'You do not have enough NEO to send.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  } else if (asset === 'Gas' && parseFloat(sendAmount.value) > gasBalance) {
    dispatch(sendEvent(false, 'You do not have enough GAS to send.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  } else if (parseFloat(sendAmount.value) < 0) { // check for negative asset
    dispatch(sendEvent(false, 'You cannot send negative amounts of an asset.'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  }
  return true
}

// open confirm pane and validate fields
const openAndValidate = (dispatch, neoBalance, gasBalance, asset) => {
  if (validateForm(dispatch, neoBalance, gasBalance, asset) === true) {
    dispatch(togglePane('confirmPane'))
  }
}

// perform send transaction
const sendTransaction = (dispatch, net, selfAddress, wif, asset, neoBalance, gasBalance) => {
  let assetName
  if (asset === 'Neo') {
    assetName = 'NEO'
  } else if (asset === 'Gas') {
    assetName = 'GAS'
  } else {
    dispatch(sendEvent(false, 'That asset is not Neo or Gas'))
    setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    return false
  }
  // validate fields again for good measure (might have changed?)
  if (validateForm(dispatch, neoBalance, gasBalance, asset) === true) {
    let sendAsset = {}
    sendAsset[assetName] = sendAmount.value
    dispatch(sendEvent(true, 'Processing...'))
    log(net, 'SEND', selfAddress, {to: sendAddress.value, asset: asset, amount: sendAmount.value})
    ledgerNanoS_doSendAsset(net, sendAddress.value, wif, sendAsset).then((response) => {
      if (response.result === undefined || response.result === false) {
        dispatch(sendEvent(false, 'Transaction failed!'))
      } else {
        console.log(response.result)
        dispatch(sendEvent(true, 'Transaction complete! Your balance will automatically update when the blockchain has processed it.'))
      }
      setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    }).catch((e) => {
      dispatch(sendEvent(false, 'Transaction failed!'))
      setTimeout(() => dispatch(clearTransactionEvent()), 5000)
    })
  }
  // close confirm pane and clear fields
  dispatch(togglePane('confirmPane'))
  sendAddress.value = ''
  sendAmount.value = ''
  confirmButton.blur()
}

let Send = ({ dispatch, wif, address, status, neo, gas, net, confirmPane, selectedAsset }) => {
  let confirmPaneClosed
  if (confirmPane) {
    confirmPaneClosed = '100%'
  } else {
    confirmPaneClosed = '69%'
  }
  return (<SplitPane className='confirmSplit' split='horizontal' size={confirmPaneClosed} allowResize={false}>
    <div id='sendPane'>
      <div id='sendAddress'>
        <input placeholder='Where to send the asset (address)' ref={node => { sendAddress = node }} />
      </div>
      <div id='sendAmount'>
        <input placeholder='Amount' ref={node => { sendAmount = node }} />
      </div>
      <button id='sendAsset' data-tip data-for='assetTip' onClick={() => dispatch(toggleAsset())}>{selectedAsset}</button>
      <ReactTooltip class='solidTip' id='assetTip' place='bottom' type='dark' effect='solid'>
        <span>Toggle NEO / GAS</span>
      </ReactTooltip>
      <button id='doSend' onClick={() => openAndValidate(dispatch, neo, gas, selectedAsset)}>Send Asset</button>
    </div>
    <div id='confirmPane' onClick={() => sendTransaction(dispatch, net, address, wif, selectedAsset, neo, gas)}>
      <button ref={node => { confirmButton = node }}>Confirm Transaction</button>
    </div>
  </SplitPane>)
}

const mapStateToProps = (state) => ({
  wif: state.account.wif,
  address: state.account.address,
  net: state.metadata.network,
  neo: state.wallet.Neo,
  gas: state.wallet.Gas,
  selectedAsset: state.transactions.selectedAsset,
  confirmPane: state.dashboard.confirmPane
})

Send.propTypes = {
  dispatch: PropTypes.func.isRequired,
  address: PropTypes.string,
  wif: PropTypes.string,
  neo: PropTypes.number,
  net: PropTypes.string,
  gas: PropTypes.number,
  confirmPane: PropTypes.bool,
  selectedAsset: PropTypes.string,
  status: PropTypes.string
}

Send = connect(mapStateToProps)(Send)

export default Send
