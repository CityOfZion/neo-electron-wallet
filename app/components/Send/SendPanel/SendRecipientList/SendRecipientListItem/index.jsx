// @flow

import React, { Component } from 'react'

import SelectInput from '../../../../Inputs/SelectInput'
import NumberInput from '../../../../Inputs/NumberInput'
import TextInput from '../../../../Inputs/TextInput'

import TrashCanIcon from '../../../../../assets/icons/delete.svg'

import styles from '../SendRecipientList.scss'

type Props = {
  asset: string,
  amount: numbers,
  address: string,
  note: string,
  index: string,
  sendableAssets: array,
  removeRow: Function,
  updateRowField: Function
}

class SendRecipientListItem extends Component<Props> {
  handleFieldChange = e => {
    const { index, updateRowField } = this.props
    const { name, value } = e.currentTarget

    updateRowField(index, name, value)
  }

  handleMaxClick = e => {
    const { index, updateRowField } = this.props

    updateRowField(index, 'amount', 20) // temporary hardcoded max
  }

  handleDeleteRow = () => {
    const { index, removeRow } = this.props

    removeRow(index)
  }

  createAssetList = () =>
    Object.keys(this.props.sendableAssets).map(asset => asset)

  render() {
    const { index, address, amount, note, asset } = this.props

    return (
      <li className={styles.sendRecipientListItem}>
        <div className={styles.rowNumber}>{`0${index + 1}`}</div>
        <div className={styles.asset}>
          <SelectInput
            value={asset}
            name="asset"
            onChange={this.handleFieldChange}
            items={this.createAssetList()}
            customChangeEvent
            disabled
          />
        </div>
        <div className={styles.amount}>
          <NumberInput
            max={20}
            value={amount}
            name="amount"
            onChange={this.handleFieldChange}
            customChangeEvent
            handleMaxClick={this.handleMaxClick}
          />
        </div>
        <div className={styles.address}>
          <SelectInput
            placeholder="Add wallet or select contact"
            value={address}
            name="address"
            onChange={this.handleFieldChange}
            customChangeEvent
          />
        </div>
        <div className={styles.reference}>
          <TextInput
            placeholder="Add a note"
            value={note}
            name="note"
            onChange={this.handleFieldChange}
          />
        </div>
        <div className={styles.delete}>
          <button
            type="button"
            className={styles.deleteButton}
            onClick={this.handleDeleteRow}
          >
            <TrashCanIcon />
          </button>
        </div>
      </li>
    )
  }
}

export default SendRecipientListItem
