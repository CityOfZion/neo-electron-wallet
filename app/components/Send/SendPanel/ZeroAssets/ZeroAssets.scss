.zeroAssets {
    display: flex;
    padding: 40px 30px;
    align-items: center;

    svg {
        height: 150px;
        width: 150px;
        min-width: 150px;
    }

    .zeroAssetsHeaderText {
        font-weight: 300;
        font-size: 2em;
        margin: 0;
    }
}