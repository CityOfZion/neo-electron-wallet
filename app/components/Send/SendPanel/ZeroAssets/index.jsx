import React from 'react'

import NotificationIcon from '../../../../assets/icons/notification.svg'

import styles from './ZeroAssets.scss'

const ZeroAssets = () => (
  <section className={styles.zeroAssets}>
    <NotificationIcon />
    <h1 className={styles.zeroAssetsHeaderText}>You have no assets to send.</h1>
  </section>
)

export default ZeroAssets
