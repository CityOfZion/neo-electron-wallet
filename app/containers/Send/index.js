import { compose } from 'recompose'
import { filter, values, omit } from 'lodash'
import Send from './Send'

import withNetworkData from '../../hocs/withNetworkData'
import withAuthData from '../../hocs/withAuthData'
import withBalancesData from '../../hocs/withBalancesData'
import withCurrencyData from '../../hocs/withCurrencyData'
import withFilteredTokensData from '../../hocs/withFilteredTokensData'

const filterSendableAssets = balances => {
  const sendableAssets = {}
  if (Number(balances.NEO > 0)) {
    sendableAssets.NEO = { symbol: 'NEO', balance: balances.NEO }
  }

  if (Number(balances.GAS > 0)) {
    sendableAssets.GAS = { symbol: 'GAS', balance: balances.NEO }
  }

  values(omit(balances, 'NEO', 'GAS'))
    .filter(token => token.balance > 0)
    .forEach(token => {
      sendableAssets[token.symbol] = {
        symbol: token.symbol,
        balance: token.balance
      }
    })

  return sendableAssets
}

const mapBalanceDataToProps = balances => ({
  NEO: balances.NEO,
  GAS: balances.GAS,
  tokenBalances: values(omit(balances, 'NEO', 'GAS')),
  sendableAssets: filterSendableAssets(balances)
})

export default compose(
  withBalancesData(mapBalanceDataToProps),
  withCurrencyData('currencyCode'),
  withNetworkData(),
  withAuthData(),
  withFilteredTokensData()
)(Send)
