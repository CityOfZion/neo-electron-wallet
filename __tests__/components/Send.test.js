import React from 'react'
import { mount } from 'enzyme'
import { uniqueId } from 'lodash-es'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux'
import { progressValues } from 'spunky'

import Send from '../../app/containers/Send/Send'
import ZeroAssets from '../../app/components/ZeroAssets/ZeroAssets'
import SendRecipientListItem from '../../app/components/Send/SendPanel/SendRecipientList/SendRecipientListItem'
import PanelHeaderButton from '../../app/components/PanelHeaderButton/PanelHeaderButton'
import Button from '../../app/components/Button'
import IntlWrapper from '../../app/components/Root/IntlWrapper'
import { DEFAULT_LANGUAGE } from '../../app/core/constants'

const { LOADED, LOADING } = progressValues

const initialState = {
  spunky: {
    settings: {
      batch: false,
      progress: LOADED,
      loadedCount: 1,
      data: {
        language: DEFAULT_LANGUAGE,
      },
    },
  },
}

const setup = props => {
  const store = configureStore()(initialState)
  const wrapper = mount(
    <Provider store={store}>
      <IntlWrapper>
        <Send
          {...props}
          sendableAssets={{ NEO: { balance: 5, symbol: 'NEO' } }}
          prices={{ NEO: 38 }}
          contacts={{ NeoFriend: 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr' }}
          currencyCode="usd"
          shouldRenderHeaderBar={false}
        />{' '}
      </IntlWrapper>
    </Provider>,
  )
  return wrapper
}

const createAsset = (name, id) => ({
  asset: name,
  address: '',
  note: '',
  amount: 0,
  max: 10,
  id,
  errors: {},
})

describe('Send', () => {
  test('It shows ZeroAssets component when no sendable assets is present', () => {
    const wrapper = setup()

    expect(wrapper.find(ZeroAssets)).toBeTruthy()
  })

  test('It does not add a row when you click Add Recipient button and without details for the first row being filled in', () => {
    const wrapper = setup()
    wrapper
      .find(PanelHeaderButton)
      .at(2)
      .simulate('click')
    expect(wrapper.find(Send).instance().state.sendRowDetails.length).toBe(1)
    expect(wrapper.find(SendRecipientListItem).children().length).toBe(1)
  })

  test('It does add a row when you click Add Recipient button and details are filled in', () => {
    const wrapper = setup()

    const sendRowDetails = []
    const asset = createAsset('GAS', 1)
    asset.amount = 1.523
    asset.address = 'ARU4Sw9yyqgfjxfqF1TNwWHHFvLbAVdTj1'
    asset.id = uniqueId()
    wrapper.find(Send).setState({
      sendRowDetails: [asset],
    })

    wrapper
      .find(PanelHeaderButton)
      .at(2)
      .simulate('click')
    expect(wrapper.find(Send).instance().state.sendRowDetails.length).toBe(2)
    expect(wrapper.find(SendRecipientListItem).children().length).toBe(2)
  })

  test('It deletes the correct row when you click the delete icon for that row', () => {
    const wrapper = setup()

    const sendRowDetails = []
    sendRowDetails.push(
      createAsset('NEO', 1),
      createAsset('GAS', 2),
      createAsset('RPX', 3),
    )

    wrapper.find(Send).setState({ sendRowDetails })
    wrapper
      .find('.deleteButton')
      .first()
      .simulate('click')

    const wrapperRows = wrapper.find(Send).instance().state.sendRowDetails

    expect(wrapperRows.length).toBe(2)
    expect(wrapperRows[0].asset).toBe('GAS')
    expect(wrapperRows[1].asset).toBe('RPX')
  })

  // TODO: update this test so that valid values are added
  test('.sendPanelHeaderButton is disabled until details are filled out', () => {
    const wrapper = setup()

    for (let i = 0; i <= 10; i++) {
      wrapper
        .find(PanelHeaderButton)
        .at(2)
        .simulate('click')
    }

    expect(wrapper.find(Send).instance().state.sendRowDetails.length).toBe(1)
    expect(wrapper.find(SendRecipientListItem).children().length).toBe(1)
  })

  test('There is no trash icon when only one row is present', () => {
    const wrapper = setup()
    expect(wrapper.find('.deleteButton')).toHaveLength(0)
  })

  test('It validates and sets error in state when you submit empty inputs', () => {
    const wrapper = setup()

    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors

    expect(errors.amount).toBe('Can not send 0 NEO.')
    expect(errors.address).toBe('You need to specify a valid NEO address.')
  })

  test('It does not allow negative amounts', () => {
    const wrapper = setup()

    const asset = createAsset('NEO', 1)
    asset.amount = -5
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors
    expect(errors.amount).toBe('You cannot send negative amounts of NEO.')
  })

  test('It does not allow fractional amounts of NEO', () => {
    const wrapper = setup()

    const asset = createAsset('NEO', 1)
    asset.amount = 1.5
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors
    expect(errors.amount).toBe('You cannot send fractional amounts of NEO.')
  })

  test('It fails when trying to send more than you have', () => {
    const wrapper = setup()

    const asset = createAsset('NEO', 1)
    asset.amount = 100
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors
    expect(errors.amount).toBe(
      'You do not have enough balance to send 100 NEO.',
    )
  })

  test('It does allow fractional amounts of other assets', () => {
    const wrapper = setup()

    const asset = createAsset('GAS', 1)
    asset.amount = 1.523
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors
    expect(errors.amount).toBe(undefined)
  })

  test('It does not allow blacklisted addresses', () => {
    const wrapper = setup()

    const asset = createAsset('GAS', 1)
    asset.amount = 1.523
    asset.address = 'ARU4Sw9yyqgfjxfqF1TNwWHHFvLbAVdTj1'
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    setTimeout(() => {
      const errors = wrapper.find(Send).instance().state.sendRowDetails[0]
        .errors

      expect(errors.address).toBe(
        'Address is blacklisted. This is a known phishing address.',
      )
    }, 0)
  })

  test('It does allow you to send to your own address', () => {
    const wrapper = setup({ address: 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr' })

    const asset = createAsset('GAS', 1)
    asset.amount = 1.523
    asset.address = 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr'
    wrapper.find(Send).setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const errors = wrapper.find(Send).instance().state.sendRowDetails[0].errors

    expect(errors.address).toBe(undefined)
  })

  test('It correctly sets max value', () => {
    const wrapper = setup()

    wrapper
      .find('.maxButton')
      .at(1)
      .simulate('click')
    expect(wrapper.find(Send).instance().state.sendRowDetails[0].amount).toBe(
      '5',
    )
  })

  test('It proceeds to the next step with valid input', () => {
    const wrapper = setup()

    const asset = createAsset('NEO', 1)
    asset.amount = 1
    asset.address = 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr'
    wrapper.setState({ sendRowDetails: [asset] })
    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    setTimeout(
      () =>
        expect(wrapper.find(Send).instance().state.showConfirmSend).toBe(true),
      0,
    )
  })

  test('It calls sendTransaction when you click confirm and send', () => {
    const wrapper = setup({ sendTransaction: jest.fn(() => Promise.resolve()) })

    const asset = createAsset('NEO', 1)
    asset.amount = 1
    asset.address = 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr'
    wrapper.setState({ sendRowDetails: [asset] })

    wrapper
      .find('#send-assets')
      .find(Button)
      .props()
      .onClick()

    const sendTransaction = wrapper.find(Send).instance().props.sendTransaction

    setTimeout(() => {
      expect(sendTransaction).toHaveBeenCalledTimes(1)
      expect(sendTransaction).toHaveBeenCalledWith([
        {
          address: 'AMKxqiSSLR89wLVEk5CoGRjKHRrmrR8bDr',
          amount: 1,
          note: '',
          symbol: 'NEO',
        },
      ])
    }, 0)
  })
})
