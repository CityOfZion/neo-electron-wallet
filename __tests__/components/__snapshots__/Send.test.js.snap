// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send renders without crashing 1`] = `
<Send
  confirmPane={true}
  gas={1}
  neo={5}
  selectedAsset="Neo"
  sendTransaction={[Function]}
  showErrorNotification={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  toggleAsset={[Function]}
  togglePane={[Function]}
/>
`;
